### PR TITLE
Abstract out site-wide variables into a single include

### PR DESCRIPTION
**Pull request by andrewdavidwong**
Known issues:
- `variables.liquid` is itself still repetitive. I tried a `for` loop instead of repeating myself but couldn't get it to work.
- This increases build time considerably. I'm sure it can be optimized in some way.

CC: @marmarek, @tokideveloper, @maiska

### CODE REVIEW

**Comment by tokideveloper** [comment]
I tested a bunch of pages and it seems to work. Nice!

The build time is now around three times as long. That's okay for me.

BTW: HTML comments like
```html
<!-- Site-wide variables -->
```
appear in the resulting HTML file. So, you could use Liquid comments instead:
```liquid
{% comment %} Site-wide variables {% endcomment %}
```

File: _includes/variables.liquid [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  {% endif %}
+  <!-- Style guide page -->
+  {% assign style-guide-page = site.data.translation[page.lang].style-guide-page %}
+  {% if style-guide-page == nil or style-guide-page.size == 0 %}
+    {% assign style-guide-page = site.data.hcl %}
+  {% endif %}
+{% endif %}
+<!-- Doc link -->
+{% assign doc_link = lang | append: "/doc/" %}
+{% assign doc_count = site.doc | concat: site.qubes-translated | where:'permalink', doc_link %}
```
Comment (marmarek) [inline _includes/variables.liquid:67]: I _guess_ this, and the `concat` below significantly contributes to the increased build time. First, for looking for translated pages, it should be enough to use just `site.qubes-translated` (without merging with `site.doc`). But then, I wonder if in this very case isn't enough to check `site.data.translation[page.lang].doc-index`, as this is where the index is generated from? You can probably move it above, where you set `doc-index` variable.

File: _includes/variables.liquid [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+{% if doc_count.size == 0 %}
+  {% assign doc_link = "/doc/" %}
+{% endif %}
+<!-- Team link -->
+{% assign team_link = lang | append: "/team/" %}
+{% assign team_count = site.pages | concat: site.qubes-translated | where:'permalink', team_link %}
```
Comment (marmarek) [inline _includes/variables.liquid:73]: For this one again, searching just `site.qubes-translated` should be enough. But maybe also `site.data.translation[page.lang].team`? Or maybe even set it together with `team-page` variable above?

**Comment by andrewdavidwong** [comment]
> BTW: HTML comments like
> 
> ```
> 
> ```
> 
> appear in the resulting HTML file. So, you could use Liquid comments instead:
> 
> ```
> {% comment %} Site-wide variables {% endcomment %}
> ```

IMHO, liquid comments make the source file too cumbersome to read, since there's no syntax highlighting, and it looks exactly like the actual content when visually scanning the file. Since the HTML comments are harmless, I think it's worth just leaving them in, but I'm open to being convinced otherwise.

File: _includes/variables.liquid (hunk 1 above)
Comment (andrewdavidwong) [reply _includes/variables.liquid:67]: I actually have no idea. I simply copied this line verbatim from the existing code (i.e., abstracted it out without modifying it).

I will try to implement your suggestion, but I'm not set up to test the translated website.

File: _includes/variables.liquid (hunk 2 above)
Comment (andrewdavidwong) [reply _includes/variables.liquid:73]: (Repeat comment, also applies here:)

I actually have no idea. I simply copied this line verbatim from the existing code (i.e., abstracted it out without modifying it).

I will try to implement your suggestion, but I'm not set up to test the translated website.

File: _includes/variables.liquid (hunk 1 above)
Comment (andrewdavidwong) [reply _includes/variables.liquid]: Done: https://github.com/QubesOS/qubesos.github.io/pull/178/commits/613c3ddf452466d6bef471fb5f546efd6f0be96d

Is this what you had in mind?

File: _includes/variables.liquid (hunk 2 above)
Comment (andrewdavidwong) [reply _includes/variables.liquid]: Done: https://github.com/QubesOS/qubesos.github.io/pull/178/commits/59a5d15deb5813683510995ec58ffa43393d5123

Is this what you had in mind?

**Comment by andrewdavidwong** [comment]
After implementing @marmarek's suggested changes regarding the assignment of `doc_link` and `team_link`, build time is back to normal (prior to this commit) levels!

**Comment by andrewdavidwong** [comment]
> Since the HTML comments are harmless, I think it's worth just leaving them in, but I'm open to being convinced otherwise.

Also, this is yet another reason to minify our output (along with https://github.com/QubesOS/qubesos.github.io/pull/177).

File: _includes/variables.liquid (hunk 2 above)
Comment (marmarek) [reply _includes/variables.liquid]: Yes, exactly.

**Comment by marmarek** [comment]
PipelineRetry

**Comment by tokideveloper** [comment]
> After implementing @marmarek's suggested changes regarding the assignment of `doc_link` and `team_link`, build time is back to normal (prior to this commit) levels!

I can confirm this. Also, I tested some of the non-translated pages and they seem to be good (I'm also not set up to test the translated website).

**Comment by andrewdavidwong** [comment]
> > BTW: HTML comments like
> > ```
> > 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > appear in the resulting HTML file. So, you could use Liquid comments instead:
> > ```
> > {% comment %} Site-wide variables {% endcomment %}
> > ```
> 
> IMHO, liquid comments make the source file too cumbersome to read, since there's no syntax highlighting, and it looks exactly like the actual content when visually scanning the file. Since the HTML comments are harmless, I think it's worth just leaving them in, but I'm open to being convinced otherwise.

Maybe a good compromise would be to rename this file to `variables.html`, since it contains HTML comments. That might be easier, anyway, since we don't have any other `.liquid` files. People might forget to use the `.liquid` extension and use `.html` out of habit.

**Comment by andrewdavidwong** [comment]
@marmarek, I don't see any actual failures in the CI log, yet it says it failed.

File: _includes/variables.liquid (hunk 1 above)
Comment (andrewdavidwong) [reply _includes/variables.liquid]: Inferring https://github.com/QubesOS/qubesos.github.io/pull/178#discussion_r657458463 also applies here. Resolving conversation.

**Comment by andrewdavidwong** [comment]
> @marmarek, I don't see any actual failures in the CI log, yet it says it failed.

Ok, the new CI log lists a lot of failures. Looks like something went wrong with the doc index...

**Comment by marmarek** [comment]
I have downloaded the built site (you can do that by clicking "browse" on the right). I'm looking at _doc/index.html and I see few issues:
 - `<!-- Variables -->` comment appears 6 times - I guess variables.html indeed got included that many times
 - the whole doc index is gone:
 ```
  <article id="doc-content" class="post-content">

<!-- BEGIN CONTENT -->


<h1>Documentation</h1>


<!-- Check whether the current page's permalink ends in a specific word.
     If it does, include content specific to that page. -->





<!-- END CONTENT -->
<!-- Closing HTML for doc content -->
  </article>
```

**Comment by andrewdavidwong** [comment]
I have no idea why this is happening.

Apparently I handled [this merge commit](https://github.com/QubesOS/qubesos.github.io/commit/c7ebd2b63693a78c41b87b540fdef4e624f99634) incorrectly, but I have no idea what I should have done differently.

**Comment by marmarek** [comment]
`doc_link` variable is not set in the English part of variables.liquid

**Comment by marmarek** [comment]
Maybe it should be outside of that conditional part? It uses `lang` variable already, so should work in both cases the same.

**Comment by andrewdavidwong** [comment]
> `doc_link` variable is not set in the English part of variables.liquid

Ahhh, I see. Thank you.

> Maybe it should be outside of that conditional part? It uses `lang` variable already, so should work in both cases the same.

Ok, sounds good.

**Comment by andrewdavidwong** [comment]
Ah, same thing with the team link. Fixing.

**Comment by andrewdavidwong** [comment]
Looks like we're all good now! Merging.

**Comment by tokideveloper** [comment]
> I have downloaded the built site (you can do that by clicking "browse" on the right). I'm looking at _site/doc/index.html and I see few issues:
> 
>     * `<!-- Variables -->` comment appears 6 times - I guess variables.html indeed got included that many times

I now see this on the official website. Has this been forgotten or is it intentionally?

Shouldn't it be sufficient to include the variables once in `_includes/head.html`?

I could open a new pull request for that.

**Comment by tokideveloper** [comment]
Being on the very current [commit](https://github.com/QubesOS/qubesos.github.io/commit/fec15dab4ed6040ba87bc835a177c57129753c9e), I get two build warnings by Jekyll:

> Build Warning: Layout 'doc-full' requested in _doc/developer/services/admin-api.md does not exist.
> Build Warning: Layout 'doc-index' requested in _doc/doc.md does not exist.

So, what is wrong here? Should the mentioned files have other layouts? Or was it wrong to delete the mentioned layouts?

**Comment by tokideveloper** [comment]
t

> Being on the very current [commit](https://github.com/QubesOS/qubesos.github.io/commit/fec15dab4ed6040ba87bc835a177c57129753c9e), I get two build warnings by Jekyll:
> 
> > Build Warning: Layout 'doc-full' requested in _doc/developer/services/admin-api.md does not exist.
> > Build Warning: Layout 'doc-index' requested in _doc/doc.md does not exist.
> 
> So, what is wrong here? Should the mentioned files have other layouts? Or was it wrong to delete the mentioned layouts?

Oh, wait, I think the mistake is on my side. I didn't check out the correct submodules/commits, I guess.

**Comment by tokideveloper** [comment]
> Shouldn't it be sufficient to include the variables once in `_includes/head.html`?

I tried it out now and it's a bad idea: Almost everything vanishes. However, it "shows" (in a sense) the snippets that are missing in the data files for translations.

**Comment by andrewdavidwong** [comment]
> > I have downloaded the built site (you can do that by clicking "browse" on the right). I'm looking at _site/doc/index.html and I see few issues:
> > ```
> > * `<!-- Variables -->` comment appears 6 times - I guess variables.html indeed got included that many times
> > ```
> 
> I now see this on the official website. Has this been forgotten or is it intentionally?
> 

Not forgotten. Since `{% variables.html %}` itself is included once at the top of many files in `_includes`, and since more than one such file can potentially be included in a page, I don't see how to avoid having `{% variables.html %}` end up included more than once on a page.

As I said above, the comments themselves are harmless (https://github.com/QubesOS/qubesos.github.io/pull/178#issuecomment-867116941), and it's better to remove them by minifying our output (https://github.com/QubesOS/qubesos.github.io/pull/178#issuecomment-867137865, https://github.com/QubesOS/qubes-issues/issues/6731).

Of course, another option is simply to remove the comments entirely, but I don't think we should uncomment our code just to avoid seeing comments in the unminified output. That would seem perverse.

**Comment by andrewdavidwong** [comment]
> > > I have downloaded the built site (you can do that by clicking "browse" on the right). I'm looking at _site/doc/index.html and I see few issues:
> > > ```
> > > * `<!-- Variables -->` comment appears 6 times - I guess variables.html indeed got included that many times
> > > ```
> > 
> > 
> > I now see this on the official website. Has this been forgotten or is it intentionally?
> 
> Not forgotten. Since `{% variables.html %}` itself is included once at the top of many files in `_includes`, and since more than one such file can potentially be included in a page, I don't see how to avoid having `{% variables.html %}` end up included more than once on a page.
> 
> As I said above, the comments themselves are harmless ([#178 (comment)](https://github.com/QubesOS/qubesos.github.io/pull/178#issuecomment-867116941)), and it's better to remove them by minifying our output ([#178 (comment)](https://github.com/QubesOS/qubesos.github.io/pull/178#issuecomment-867137865), [QubesOS/qubes-issues#6731](https://github.com/QubesOS/qubes-issues/issues/6731)).
> 
> Of course, another option is simply to remove the comments entirely, but I don't think we should uncomment our code just to avoid seeing comments in the unminified output. That would seem perverse.

BTW, it's not just from `variables.html`. If you look at the source of any page on the live website, there are many comments from other includes, as well. So, unless we rip out all of the comments I just added, which (as I mentioned above) would be perverse, we're going to have these extra HTML comments that apply to the source code but not to the HTML output until we start minifying.

**Comment by marmarek** [comment]
This comment itself is rather harmless. But it shows the `variables.html` file is included several times, which may have negative side effects. At the very least, it may make the build slower (which is not a big problem).

**Comment by andrewdavidwong** [comment]
> This comment itself is rather harmless. But it shows the `variables.html` file is included several times, which may have negative side effects. At the very least, it may make the build slower (which is not a big problem).

Right, as I said above (https://github.com/QubesOS/qubesos.github.io/pull/178#issuecomment-867532178): Since `{% variables.html %}` itself is included once at the top of many files in `_includes`, and since more than one such file can potentially be included in a page, I don't see how to avoid having `{% variables.html %}` end up included more than once on a page.

**Comment by andrewdavidwong** [comment]
> > This comment itself is rather harmless. But it shows the `variables.html` file is included several times, which may have negative side effects. At the very least, it may make the build slower (which is not a big problem).
> 
> Right, as I said above ([#178 (comment)](https://github.com/QubesOS/qubesos.github.io/pull/178#issuecomment-867532178)): Since `variables.html` itself is included once at the top of many files in `_includes`, and since more than one such file can potentially be included in a page, I don't see how to avoid having `variables.html` end up included more than once on a page.

However, as I also mentioned above (https://github.com/QubesOS/qubesos.github.io/pull/178#issuecomment-867137266) the actual build time, in practice, has not increased compared to what it was before.

**Comment by marmarek** [comment]
As long as `variables.html` is only setting new variables, it should be okay-ish. Once it starts modifying existing ones (like, appending something to an existing var), it will be a problem - the operation will be done twice (or thrice etc).
The idea of including `variables.html` once at some common, early file (or maybe just include it in layouts but not _includes?) should work, but as @tokideveloper tested, it isn't trivial to get it right.

I don't think we need to fix it now. But IMO we will need it one day. Maybe create an issue with low priority?

**Comment by andrewdavidwong** [comment]
> As long as `variables.html` is only setting new variables, it should be okay-ish. Once it starts modifying existing ones (like, appending something to an existing var), it will be a problem - the operation will be done twice (or thrice etc).
> The idea of including `variables.html` once at some common, early file (or maybe just include it in layouts but not _includes?) should work, but as @tokideveloper tested, it isn't trivial to get it right.
> 
> I don't think we need to fix it now. But IMO we will need it one day. Maybe create an issue with low priority?

Done: https://github.com/QubesOS/qubes-issues/issues/6734

The layout idea is a good one. I will experiment with that.

**Comment by tokideveloper** [comment]
Maybe we could leave all the includings of `variables.html` there but surround the whole `variables.html` with these lines:

```liquid
{% if architecture == nil or architecture.size == 0 %}
  ...
{% endif %}
```

It can also be any other variable that is set in `variables.html` in both cases (for official site and for the translated site) but I used `architecture` in this example since its a variable that saves whole pages and I experienced that sometimes such variables lose their contents (roughly said).

**Comment by tokideveloper** [comment]
> Maybe we could leave all the includings of `variables.html` there but surround the whole `variables.html` with these lines:
> 
> ```
> {% if architecture == nil or architecture.size == 0 %}
>   ...
> {% endif %}
> ```
> 
> It can also be any other variable that is set in `variables.html` in both cases (for the official site and for the translated site) but I used `architecture` in this example since it's a variable that saves whole pages and I experienced that sometimes such variables lose their contents (roughly said).

It works! :tada:  It saves some build time (from 13 sec to 11 sec on my machine) and also the number of actual variable "updates" has been reduced.

However, I wonder if I should insert these checks not for the whole thing but for every single variable separately. What do you think?

**Comment by marmarek** [comment]
> However, I wonder if I should insert these checks not for the whole thing but for every single variable separately. What do you think?

I'd go with a single check for all. Less clutter and still should do the job.

**Comment by tokideveloper** [comment]
Okay, stay tuned for my PR.

**Comment by andrewdavidwong** [comment]
> Maybe we could leave all the includings of `variables.html` there but surround the whole `variables.html` with these lines:
> 
> ```
> {% if architecture == nil or architecture.size == 0 %}
>   ...
> {% endif %}
> ```
> 
> It can also be any other variable that is set in `variables.html` in both cases (for the official site and for the translated site) but I used `architecture` in this example since it's a variable that saves whole pages and I experienced that sometimes such variables lose their contents (roughly said).

I don't understand. `architecture` isn't even assigned until inside of this, so how is it meaningful as a wrapper around the outside?

Why does this work?

**Comment by andrewdavidwong** [comment]
> > Maybe we could leave all the includings of `variables.html` there but surround the whole `variables.html` with these lines:
> > ```
> > {% if architecture == nil or architecture.size == 0 %}
> >   ...
> > {% endif %}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It can also be any other variable that is set in `variables.html` in both cases (for the official site and for the translated site) but I used `architecture` in this example since it's a variable that saves whole pages and I experienced that sometimes such variables lose their contents (roughly said).
> 
> I don't understand. `architecture` isn't even assigned until inside of this, so how is it meaningful as a wrapper around the outside?
> 

Oh, I guess that's precisely the point? `architecture == nil` since it hasn't been assigned yet, unless all of `variables.html` has already been invoked, in which case we don't want to invoke it again? But then why the `or architecture.size == 0`?

**Comment by tokideveloper** [comment]
> > > Maybe we could leave all the includings of `variables.html` there but surround the whole `variables.html` with these lines:
> > > ```
> > > {% if architecture == nil or architecture.size == 0 %}
> > >   ...
> > > {% endif %}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > It can also be any other variable that is set in `variables.html` in both cases (for the official site and for the translated site) but I used `architecture` in this example since it's a variable that saves whole pages and I experienced that sometimes such variables lose their contents (roughly said).
> > 
> > 
> > I don't understand. `architecture` isn't even assigned until inside of this, so how is it meaningful as a wrapper around the outside?
> 
> Oh, I guess that's precisely the point? `architecture == nil` since it hasn't been assigned yet, unless all of `variables.html` has already been invoked, in which case we don't want to invoke it again?

Exactly. Sometimes, variables which save pages (like `architecture`) seem to be emptied or `nil`-led (or not set yet), so we need to re-invoke the code again.

> But then why the `or architecture.size == 0`?

Just to be on the safe side. :slightly_smiling_face:

**Comment by tokideveloper** [comment]
> Okay, stay tuned for my PR.

Opened: https://github.com/QubesOS/qubesos.github.io/pull/179/

**Pull request merged**